### PR TITLE
Brian/wise ramp

### PR DESCRIPTION
**Pull request by bweick**
```
    Deploment summary for localhost:
    deployer:                   0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
    deployer nonce:             55
    multiSig:                   0x70997970C51812dc3A010C7d01b50e0d17dc79C8
    multiSig nonce:             9
    ----------------------------------------------------------------------
    Ramp:                               0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0
    HDFCRamp:                           0x0B306BF915C4d645ff596e518fAf3F9669b97016
    VenmoRampV2:                        0xa85233C63b9Ee964Add6F2cffe00Fd84eb32338f
    GarantiRamp:                        0x9E545E3C0baAB3E08CdfD552C960A1050f373042
    WiseRamp:                           0x0E801D84Fa97b50751Dbf25036d067dCf18858bF
    VenmoRegistrationProcessor:         0x5FC8d32690cc91D4c39d9d3abcBD16989F875707
    VenmoSendProcessor:                 0x0165878A594ca255338adfa4d48449f69242Eb8F
    VenmoRegistrationProcessorV2:       0x7a2088a1bFc9d81c55368AE168C2C02570cB814F
    VenmoSendProcessorV2:               0x09635F643e140090A9A8Dcd712eD6285858ceBef
    VenmoKeyHashAdapter:                0xCf7Ed3AccA5a467e9e704C703E8D87F634fB0Fc9
    VenmoKeyHashAdapterV2:              0x4A679253410272dd5232B3Ff7cF5dbB88f295319
    HDFCRegistrationProcessor:          0x9A9f2CCfdE556A7E9Ff0848998Aa4a0CFD8863AE
    HDFCSendProcessor:                  0x68B1D87F95878fE05B998F19b66F4baba5De1aed
    HDFCKeyHashAdapter:                 0x959922bE3CAee4b8Cd9a407cc3ac1C251C2007B1
    GarantiRegistrationProcessor:       0x851356ae760d987E095750cCeb3bC6014560891C
    GarantiSendProcessor:               0xf5059a5D33d5853360D16C683c16e67980206f36
    GarantiKeyHashAdapter:              0xa82fF9aFd8f496c3d6ac40E2a0F282E47488CFc9
    WiseAccountRegistry:                0x8f86403A4DE0BB5791fa46B8e795C547942fE4Cf
    WiseAccountRegistrationProcessor:   0x9d4454B023096f34B160D6B654540c56A1F81688
    WiseOffRamperRegistrationProcessor: 0x5eb3Bc0a489C5A8288765d2336659EbCA68FCd00
    WiseSendProcessor:                  0x36C02dA8a0983159322a80FFE9F24b1acfF8B570
    NullifierRegistry:                  0xDc64a140Aa3E981100a9becA4E685f962f0cF6C9
    USDC:                               0x5FbDB2315678afecb367f032d93F642f64180aa3
    Poseidon3:                          0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    Poseidon6:                          0x9A676e781A523b5d0C0e43731313A708CB607508
    ```

### CODE REVIEW

File: contracts/contracts/ramps/wise/WiseRamp.sol [hunk 1]
```diff
@@ -0,0 +1,856 @@
...
+
+pragma solidity ^0.8.18;
+
+contract WiseRamp is Ownable {
+
+    using Bytes32ArrayUtils for bytes32[];
+    using Uint256ArrayUtils for uint256[];
+
+    /* ============ Events ============ */
+    event AccountRegistered(address indexed accountOwner, bytes32 indexed onRampId, bytes32 indexed wiseTagHash);
```
Comment (0xSachinK) [inline contracts/contracts/ramps/wise/WiseRamp.sol:21]: Think we should still call onRampId as just id. because it would be held by everyone in the system, on-rampers and off-rampers alike. That is true for wise. Although not sure how that would scale beyond wise.

File: contracts/contracts/ramps/wise/interfaces/ITLSData.sol [hunk 2]
```diff
@@ -0,0 +1,11 @@
+//SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.18;
+
+interface ITLSData {
+    struct TLSParams {
```
Comment (0xSachinK) [inline contracts/contracts/ramps/wise/interfaces/ITLSData.sol:6]: This is a bit opinionated but I would break TLSParams to endpoint and headers. Or rather, endpoint, host and headers. Endpoint and host should be constant for a given processor. Where as, headers might change. And data would definitely change.

**Review by 0xSachinK — COMMENTED** [review]
Dropped a few high-level opinionated suggestions.

File: contracts/contracts/ramps/wise/WiseRegistrationProcessor.sol [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+    function setOffRamperTLSParams(ITLSData.TLSParams calldata _tlsParams) external onlyOwner {
+        offRamperTLSParams = _tlsParams;
+
+        emit OffRamperTLSParamsSet(_tlsParams);
+    }
+
+    /* ============ External Getters ============ */
+
+    function getAccountTLSParams() external view returns(ITLSData.TLSParams memory) {
```
[the remark this answers is not in the document]
Comment (bweick) [reply contracts/contracts/ramps/wise/WiseRegistrationProcessor.sol:115]: The SendProcessor doesn't need pre-defined TLSParams

File: contracts/contracts/ramps/wise/interfaces/ITLSData.sol (hunk 2 above)
Comment (bweick) [reply contracts/contracts/ramps/wise/interfaces/ITLSData.sol:6]: What do you consider headers to be? Haven't seen that term in any of the calldata. I mainly want to just group all the data required for validating TLSN transactions together (this is separate from any business logic)

File: contracts/contracts/ramps/wise/interfaces/ITLSData.sol [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+        address verifier;
```
[the remark this answers is not in the document]
Comment (bweick) [reply contracts/contracts/ramps/wise/interfaces/ITLSData.sol:7]: I'm not gonna split them out, they all make sense together as requirements to validate any one TLS proof. You need all three of those values otherwise you can't validate.

File: contracts/contracts/ramps/wise/WiseRegistrationProcessor.sol [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        ITLSData.TLSParams memory passedTLSParams = ITLSData.TLSParams({
+            verifier: address(0),                                 // Verifier not checked in validateTLSParams
+            endpoint: _proof.public_values.endpoint,
+            host: _proof.public_values.host
+        });
+
+        _validateTLSParams(accountTLSParams, passedTLSParams);
+        _validateAndAddNullifier(keccak256(abi.encode("registration", _proof.public_values.profileId)));
+
+        onRampId = bytes32(_proof.public_values.profileId.stringToUint(0));
```
[the remark this answers is not in the document]
Comment (bweick) [reply contracts/contracts/ramps/wise/WiseRegistrationProcessor.sol:67]: Compatibility with what we've previously done mainly. We also don't know that id's will always be represented as a number so a string or hash makes the most sense

File: contracts/contracts/ramps/wise/WiseSendProcessor.sol [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+        amount = _proof.public_values.amount.stringToUint(6);
+
+        // Add the buffer to build in flexibility with L2 timestamps
+        timestamp = _proof.public_values.timestamp.stringToUint(0) / 1000 + timestampBuffer;
+
+        offRamperId = bytes32(_proof.public_values.recipientId.stringToUint(0));
+        onRamperId = bytes32(_proof.public_values.senderId.stringToUint(0));
+        intentHash = bytes32(_proof.public_values.intentHash.stringToUint(0));
+        currencyId = keccak256(abi.encodePacked(_proof.public_values.currencyId));
```
[the remark this answers is not in the document]
Comment (bweick) [reply contracts/contracts/ramps/wise/WiseSendProcessor.sol:82]: `bytes32` gives us the most flexibility so I want to keep it. We aren't guaranteed to have `profileId` with only numbers for future integrations (in which case then we're essentially doing exactly what we're doing here, converting to `bytes32` representation then casting to `uint256`)

File: contracts/contracts/ramps/wise/WiseRamp.sol (hunk 1 above)
Comment (bweick) [reply contracts/contracts/ramps/wise/WiseRamp.sol:21]: Yeah one of the open questions I've been considering too. I'll make that change

File: contracts/contracts/ramps/wise/WiseRegistrationProcessor.sol [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        returns(bytes32 onRampId, bytes32 offRampId)
+    {
+        _validateOffRamperNotarySignature(_proof.public_values, _proof.proof);
+
+        ITLSData.TLSParams memory passedTLSParams = ITLSData.TLSParams({
+            verifier: address(0),                                 // Notary not checked in validateTLSParams
+            endpoint: _proof.public_values.endpoint,
+            host: _proof.public_values.host
+        });
+        ITLSData.TLSParams memory expectedTLSParams = ITLSData.TLSParams({
```
[the remark this answers is not in the document]
Comment (bweick) [reply contracts/contracts/ramps/wise/WiseRegistrationProcessor.sol:86]: I can do that

File: contracts/contracts/ramps/wise/WiseRamp.sol [hunk 8]
```diff
@@ -0,0 +1,872 @@
...
+            uint256 amount,
+            uint256 timestamp,
+            bytes32 offRamperId,
+            bytes32 onRamperId,
+            bytes32 intentHash,
+            bytes32 currencyId
+        ) = sendProcessor.processProof(
+            IWiseSendProcessor.SendProof({
+                public_values: _data,
+                expectedTLSParams: deposit.tlsParams,
```
Comment (0xSachinK) [inline contracts/contracts/ramps/wise/WiseRamp.sol:823]: Why is this required only for the send processor and not for the registration processors?

**Review by 0xSachinK — COMMENTED** [review]
Dropped another review. Mainly just around some abstractions related to TLSParams and verification key. Unless there's something different happening with the send processor, we can move the TLSParams settters, getters and verification logic into the base processor. And we should def move the key management to managed key hash adapter. 

Rest looks good to me. Haven't taken a look at tests yet.

File: contracts/contracts/ramps/wise/WiseRegistrationProcessor.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xSachinK) [reply contracts/contracts/ramps/wise/WiseRegistrationProcessor.sol]: Why? It should have one. Because it doesn't change for every wise send.

File: contracts/test/ramps/wise/wiseSendProcessor.spec.ts [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+    deployer = new DeployHelper(owner.wallet);
+
+    nullifierRegistry = await deployer.deployNullifierRegistry();
+
+    // tlsParams = {
+    //   verifierSigningKey: verifier.address,
+    // };
+
+    registrationProcessor = await deployer.deployWiseSendProcessor(
```
Comment (0xSachinK) [inline contracts/test/ramps/wise/wiseSendProcessor.spec.ts:47]: Nit: SendProcessor

File: contracts/contracts/ramps/wise/WiseSendProcessor.sol [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            _nullifierRegistry,
+            _timestampBuffer,
+            _endpoint,
+            _host
+        )
+    {}
+    
+    /* ============ External Functions ============ */
+    function processProof(
+        IWiseSendProcessor.SendProof calldata _proof
```
Comment (0xSachinK) [inline contracts/contracts/ramps/wise/WiseSendProcessor.sol:42]: All the other processors also accept proof. But in the case of send, proof also contains verifierSigningKey apart from the signature and public signals. Would advocate for keeping the passing in verifierSigningKey explicitly.

File: contracts/contracts/ramps/wise/WiseSendProcessor.sol [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        offRamperId = bytes32(_proof.public_values.recipientId.stringToUint(0));
+        onRamperId = bytes32(_proof.public_values.senderId.stringToUint(0));
+        intentHash = bytes32(_proof.public_values.intentHash.stringToUint(0));
+        currencyId = keccak256(abi.encodePacked(_proof.public_values.currencyId));
+    }
+
+    /* ============ Internal Functions ============ */
+
+    function _validateProof(
+        address _verifier,
```
Comment (0xSachinK) [inline contracts/contracts/ramps/wise/WiseSendProcessor.sol:85]: Nit: _verifierSigningKey

File: contracts/test/ramps/wise/wiseRamp.spec.ts [hunk 12]
```diff
@@ -0,0 +1,1914 @@
...
+            timeJump = ONE_DAY_IN_SECONDS.add(1).toNumber();
+          });
+
+          it("should revert", async () => {
+            await expect(subject()).to.be.revertedWith("Not enough liquidity");
+          });
+        });
+      });
+
+      describe("when the caller is on the depositor's denylist", async () => {
```
Comment (0xSachinK) [inline contracts/test/ramps/wise/wiseRamp.spec.ts:542]: Missing test cases for allow list and it's combinations with deny list

**Review by 0xSachinK — CHANGES_REQUESTED** [review]
Requested minor changes. 

Rest LGTM.

File: contracts/test/ramps/wise/wiseRamp.spec.ts (hunk 12 above)
Comment (bweick) [reply contracts/test/ramps/wise/wiseRamp.spec.ts:542]: These tests are in the WiseAccountRegistry, here just trying to show that it'll trip if `isAllowedUser` returns `false`

**Comment by 0xSachinK** [comment]
Todo: Add codecov.

**Review by 0xSachinK — APPROVED** [review]
Reviewed the changes. LGTM. Still need to add codecov.
